service/history/hsm/tree.go

```diff
@@ -82,8 +85,24 @@ type cachedMachine struct {
 	children map[Key]*Node
 	// A flag that indicates the cached machine is dirty.
 	dirty bool
-	// Outputs of all transitions in the current transaction.
-	outputs []TransitionOutputWithCount
+	// A flag to indicate the node was deleted.
+	deleted bool
+}
+
+type OperationLog struct {
+	// All transitions, including regular state changes and deletions
+	Transitions []TransitionOutputWithCount
+	// Track topmost deleted paths to filter outputs efficiently
+	DeletedPaths [][]Key
+}
```

Comment:
```suggestion
type Operation interface {
	isAnOperation()
}

type DeleteOperation struct {
	Path []Key	
}

func (DeleteOperation) isAnOperation() {}

type TransitionOperation struct {
	Path []Key	
	Output TransitionOutputWithCount
}

func (TransitionOperation) isAnOperation() {}

type OperationLog []Operation
```

Reply:
Alternatively `OperationLog` could also store the top-most deletes as an optimization but I don't even think it's necessary.

Reply:
Stored top-most deletes as an optimization